usr.bin/cmp/cmp.1
.\" Copyright (c) 1987, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd September 23, 2021
.Dt CMP 1
.Os
.Sh NAME
.Nm cmp
.Nd compare two files
.Sh SYNOPSIS
.Nm
.Op Fl l | s | x
.Op Fl bhz
.Op Fl -ignore-initial Ns Cm = Ns Ar num1 Ns Op :num2
.Op Fl -bytes Ns Cm = Ns Ar num
.Ar file1 file2
.Op Ar skip1 Op Ar skip2
.Sh DESCRIPTION
The
.Nm
utility compares two files of any type and writes the results
to the standard output.
By default,
.Nm
is silent if the files are the same; if they differ, the byte
and line number at which the first difference occurred is reported.
.Pp
Bytes and lines are numbered beginning with one.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl b , Fl -print-bytes
Print each byte when a difference is found.
.It Fl h
Do not follow symbolic links.
.It Fl i Ar num1 Ns Oo :num2 Oc , Fl -ignore-initial= Ns Ar num1 Ns Op :num2
Skip
.Ar num1
bytes from
.Ar file1 ,
and optionally skip
.Ar num2
bytes from
.Ar file2 .
If
.Ar num2
is not specified, then
.Ar num1
is applied for both
.Ar file1
and
.Ar file2 .
.It Fl l , Fl -verbose
Print the byte number (decimal) and the differing
byte values (octal) for each difference.
.It Fl n Ar num , Fl -bytes= Ns num
Only compare up to
.Ar num
bytes.
.It Fl s , Fl -silent , Fl -quiet
Print nothing for differing files; return exit
status only.
.It Fl x
Like
.Fl l
but prints in hexadecimal and using zero as index
for the first byte in the files.
.It Fl z
For regular files compare file sizes first, and fail the comparison if they
are not equal.
.El
.Pp
The optional arguments
.Ar skip1
and
.Ar skip2
are the byte offsets from the beginning of
.Ar file1
and
.Ar file2 ,
respectively, where the comparison will begin.
The offset is decimal by default, but may be expressed as a hexadecimal
or octal value by preceding it with a leading ``0x'' or ``0''.
.Pp
.Ar skip1
and
.Ar skip2
may also be specified with SI size suffixes.
.Sh EXIT STATUS
The
.Nm
utility exits with one of the following values:
.Bl -tag -width 4n
.It 0
The files are identical.
.It 1
The files are different; this includes the case
where one file is identical to the first part of
the other.
In the latter case, if the
.Fl s
option has not been specified,
.Nm
writes to standard error that EOF was reached in the shorter
file (before any differences were found).
.It >1
An error occurred.
.El
.Sh EXAMPLES
Assuming a file named
.Pa example.txt
with the following contents:
.Bd -literal -offset indent
a
b
c
.Ed
.Pp
Compare stdin with
.Pa example.txt :
.Bd -literal -offset indent
$ echo -e "a\\nb\\nc" | cmp - example.txt
.Ed
.Pp
Same as above but introducing a change in the third byte of stdin.
Show the byte number (decimal) and differing byte (octal):
.Bd -literal -offset indent
$ echo -e "a\\nR\\nc" | cmp -l - example.txt
     3 122 142
.Ed
.Pp
Compare file sizes of
.Pa example.txt
and
.Pa /boot/loader.conf
and return 1 if they are not equal.
Note that
.Fl z
can only be used with regular files:
.Bd -literal -offset indent
$ cmp -z example.txt /boot/loader.conf
example.txt /boot/loader.conf differ: size
.Ed
.Pp
Compare stdin with
.Pa example.txt
omitting the first 4 bytes from stdin and the first 2 bytes from
.Pa example.txt :
.Bd -literal -offset indent
$ echo -e "a\\nR\\nb\\nc" | cmp - example.txt 4 2
.Ed
.Sh SEE ALSO
.Xr diff 1 ,
.Xr diff3 1
.Sh STANDARDS
The
.Nm
utility is expected to be
.St -p1003.2
compatible.
The
.Fl b ,
.Fl h ,
.Fl i ,
.Fl n ,
.Fl x ,
and
.Fl z
options are extensions to the standard.
.Ar skip1
and
.Ar skip2
arguments are extensions to the standard.
.Sh HISTORY
A
.Nm
command appeared in
.At v1 .
.Sh BUGS
The phrase
.Dq SI size suffixes
above refers to the traditional power of two convention, as described in
.Xr expand_number 3 .


NEWS
Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader
Jun 16, 2011: Add tar:compat-2x option to emulate broken libarchive 2.x
   handling of pax UTF-8 headers
Apr 25, 2011: Refactor read_open() into a collection of single-item setters;
   support the old interfaces as wrappers
Apr 12, 2011: Split disk writer into separate POSIX and Windows implementations
Apr 10, 2011: Improvements to character translations on Windows.
Mar 30, 2011: More work to return errors instead of calling abort()

usr.bin/cmp/cmp.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Copyright (c) 1987, 1990, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <capsicum_helpers.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <nl_types.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <libutil.h>

#include "extern.h"

bool	bflag, lflag, sflag, xflag, zflag;

static const struct option long_opts[] =
{
	{"print-bytes",	no_argument,		NULL, 'b'},
	{"ignore-initial", required_argument,	NULL, 'i'},
	{"verbose",	no_argument,		NULL, 'l'},
	{"bytes",	required_argument,	NULL, 'n'},
	{"silent",	no_argument,		NULL, 's'},
	{"quiet",	no_argument,		NULL, 's'},
	{NULL,		no_argument,		NULL, 0}
};

#ifdef SIGINFO
volatile sig_atomic_t info;

static void
siginfo(int signo)
{

	info = signo;
}
#endif

static void usage(void) __dead2;

static bool
parse_iskipspec(char *spec, off_t *skip1, off_t *skip2)
{
	char *colon;

	colon = strchr(spec, ':');
	if (colon != NULL)
		*colon++ = '\0';

	if (expand_number(spec, skip1) < 0)
		return (false);

	if (colon != NULL)
		return (expand_number(colon, skip2) == 0);

	*skip2 = *skip1;
	return (true);
}

int
main(int argc, char *argv[])
{
	struct stat sb1, sb2;
	off_t skip1, skip2, limit;
	int ch, fd1, fd2, oflag;
	bool special;
	const char *file1, *file2;

	limit = skip1 = skip2 = 0;
	oflag = O_RDONLY;
	while ((ch = getopt_long(argc, argv, "+bhi:ln:sxz", long_opts, NULL)) != -1)
		switch (ch) {
		case 'b':		/* Print bytes */
			bflag = true;
			break;
		case 'h':		/* Don't follow symlinks */
			oflag |= O_NOFOLLOW;
			break;
		case 'i':
			if (!parse_iskipspec(optarg, &skip1, &skip2)) {
				fprintf(stderr,
				    "Invalid --ignore-initial: %s\n",
				    optarg);
				usage();
			}
			break;
		case 'l':		/* print all differences */
			lflag = true;
			break;
		case 'n':		/* Limit */
			if (expand_number(optarg, &limit) < 0 || limit < 0) {
				fprintf(stderr, "Invalid --bytes: %s\n",
				    optarg);
				usage();
			}
			break;
		case 's':		/* silent run */
			sflag = true;
			break;
		case 'x':		/* hex output */
			lflag = true;
			xflag = true;
			break;
		case 'z':		/* compare size first */
			zflag = true;
			break;
		case '?':
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	if (lflag && sflag)
		errx(ERR_EXIT, "specifying -s with -l or -x is not permitted");

	if (argc < 2 || argc > 4)
		usage();

	/* Don't limit rights on stdin since it may be one of the inputs. */
	if (caph_limit_stream(STDOUT_FILENO, CAPH_WRITE | CAPH_IGNORE_EBADF))
		err(ERR_EXIT, "unable to limit rights on stdout");
	if (caph_limit_stream(STDERR_FILENO, CAPH_WRITE | CAPH_IGNORE_EBADF))
		err(ERR_EXIT, "unable to limit rights on stderr");

	/* Backward compatibility -- handle "-" meaning stdin. */
	special = false;
	if (strcmp(file1 = argv[0], "-") == 0) {
		special = true;
		fd1 = STDIN_FILENO;
		file1 = "stdin";
	} else if ((fd1 = open(file1, oflag, 0)) < 0 && errno != EMLINK) {
		if (!sflag)
			err(ERR_EXIT, "%s", file1);
		else
			exit(ERR_EXIT);
	}
	if (strcmp(file2 = argv[1], "-") == 0) {
		if (special)
			errx(ERR_EXIT,
				"standard input may only be specified once");
		special = true;
		fd2 = STDIN_FILENO;
		file2 = "stdin";
	} else if ((fd2 = open(file2, oflag, 0)) < 0 && errno != EMLINK) {
		if (!sflag)
			err(ERR_EXIT, "%s", file2);
		else
			exit(ERR_EXIT);
	}

	if (argc > 2 && expand_number(argv[2], &skip1) < 0) {
		fprintf(stderr, "Invalid skip1: %s\n", argv[2]);
		usage();
	}

	if (argc == 4 && expand_number(argv[3], &skip2) < 0) {
		fprintf(stderr, "Invalid skip2: %s\n", argv[3]);
		usage();
	}

	if (sflag && skip1 == 0 && skip2 == 0)
		zflag = true;

	if (fd1 == -1) {
		if (fd2 == -1) {
			c_link(file1, skip1, file2, skip2, limit);
			exit(0);
		} else if (!sflag)
			errx(ERR_EXIT, "%s: Not a symbolic link", file2);
		else
			exit(ERR_EXIT);
	} else if (fd2 == -1) {
		if (!sflag)
			errx(ERR_EXIT, "%s: Not a symbolic link", file1);
		else
			exit(ERR_EXIT);
	}

	/* FD rights are limited in c_special() and c_regular(). */
	caph_cache_catpages();

	if (!special) {
		if (fstat(fd1, &sb1)) {
			if (!sflag)
				err(ERR_EXIT, "%s", file1);
			else
				exit(ERR_EXIT);
		}
		if (!S_ISREG(sb1.st_mode))
			special = true;
		else {
			if (fstat(fd2, &sb2)) {
				if (!sflag)
					err(ERR_EXIT, "%s", file2);
				else
					exit(ERR_EXIT);
			}
			if (!S_ISREG(sb2.st_mode))
				special = true;
		}
	}

#ifdef SIGINFO
	(void)signal(SIGINFO, siginfo);
#endif
	if (special)
		c_special(fd1, file1, skip1, fd2, file2, skip2, limit);
	else {
		if (zflag && sb1.st_size != sb2.st_size) {
			if (!sflag)
				(void) printf("%s %s differ: size\n",
				    file1, file2);
			exit(DIFF_EXIT);
		}
		c_regular(fd1, file1, skip1, sb1.st_size,
		    fd2, file2, skip2, sb2.st_size, limit);
	}
	exit(0);
}

static void
usage(void)
{

	(void)fprintf(stderr,
	    "usage: cmp [-l | -s | -x] [-hz] file1 file2 [skip1 [skip2]]\n");
	exit(ERR_EXIT);
}
